### PR TITLE
NEWRELIC-5563 added helper to retrieve CLM meta from a function refer…

### PR DESCRIPTION
**Pull request by bizob2828**
…ence

## Proposed Release Notes
 * Added a utility to retrieve Code Level Metrics(CLM) metadata from a function reference.

## Links
Closes NEWRELIC-5563

## Details
The library I'm using is from a company I used to work for.  This does exactly what we need but using the v8 APIs which is not nearly as expensive as making an error and calling stack getter.  it does lack column, but I reached out to the team at Contrast and I plan on PR'ing this support

### CODE REVIEW

**Comment by bizob2828** [comment]
We can either wait or land this and update once this gets merged and released https://github.com/Contrast-Security-Inc/node-fn-inspect/pull/42

File: lib/util/code-level-metrics.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+const logger = require('../logger').child({ component: 'code-level-metrics' })
+
+/**
+ * Helper to retrieve Code Level Metrics(CLM)
+ * properties from a function reference.
+ *
+ * @param {Function} fn function reference
+ * @returns {object} CLM properties
+ */
+module.exports = function getCLMMeta(fn) {
```
Comment (jmartin4563) [inline lib/util/code-level-metrics.js:16]: Not to be a buzzkill, but the agent-spec has more rules that I don't see accounted for here, specifically 

```
Agents MUST:

drop the attribute if it's over 255 characters
if code.function doesn't exist or if the function name is over 255 characters, drop the CLM attributes
if code.function exists and both code.namespace and code.filepath are over 255 characters, drop the CLM attributes
```

**Review by jmartin4563 — CHANGES_REQUESTED** [review]

File: lib/util/code-level-metrics.js (hunk 1 above)
Comment (bizob2828) [reply lib/util/code-level-metrics.js:16]: ah good catch. I was going to do this then go distracted with finding an easier solution for retrieving the metadata

File: lib/util/code-level-metrics.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  try {
+    const { funcInfo } = require('@contrast/fn-inspect')
+    const fnMeta = funcInfo(fn)
+    return {
+      'code.filepath': fnMeta.file,
+      'code.function': fnMeta.method || 'anonymous',
+      'code.lineno': fnMeta.lineNumber + 1 // line numbers start at 0 in v8 so we have to add 1 to reflect js code
```
[the remark this answers is not in the document]
Comment (bizob2828) [reply lib/util/code-level-metrics.js:23]: Doubtful

File: lib/util/code-level-metrics.js (hunk 1 above)
Comment (bizob2828) [reply lib/util/code-level-metrics.js:16]: this is built into our addAttribute, so eventually when these keys get associated with a span it'll drop if > 255 https://github.com/newrelic/node-newrelic/blob/main/lib/attributes.js#L127

Comment (bizob2828) [reply lib/util/code-level-metrics.js:16]: actually we truncate values, so i will build this to align with spec

File: lib/util/code-level-metrics.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    }
+  } catch (err) {
+    logger.info({ err }, 'Not using v8 function inspector, falling back to function name')
+    return {
+      'code.function': fn.name || 'anonymous'
```
[the remark this answers is not in the document]
Comment (bizob2828) [reply lib/util/code-level-metrics.js:28]: we're going with `<anonymous>` and `<arrow>`

File: lib/util/code-level-metrics.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    if (isValidLength(fnName, 255) && isValidLength(filePath, 255)) {
+      return {
+        'code.filepath': filePath,
+        'code.function': fnName,
+        'code.lineno': lineNumber + 1 // line numbers start at 0 in v8 so we have to add 1 to reflect js code
+      }
+    }
+  } catch (err) {
+    logger.info({ err }, 'Not using v8 function inspector, falling back to function name')
```
Comment (jmartin4563) [inline lib/util/code-level-metrics.js:42]: Should this be `info` or `trace`? I feel like it would pollute the logs at `info`

**Review by jmartin4563 — DISMISSED** [review]
I have one thought about the log level. Also, let me know if you want to update to the new version of the v8 lib in here, and I'll re-review.

File: lib/util/code-level-metrics.js (hunk 4 above)
Comment (bizob2828) [reply lib/util/code-level-metrics.js]: good point, changed to infoOnce

**Review by jmartin4563 — APPROVED** [review]

**Pull request merged**